File: 0410-split-array-largest-sum/0410-split-array-largest-sum.java
class Solution {
    public int splitArray(int[] nums, int k) {
        if(nums == null || nums.length < k || nums.length == 0){
            return -1;
        }

        int low = 0;
        int high = 0;

        for(int num: nums){
            low = Math.max(num, low);
            high += num;
        }

        while(low <= high){
            int mid = low + (high - low)/2;
            int subArrays = isPossible(nums,mid);
            if(subArrays > k){
                low = mid+1;
            } else{
                high = mid-1;
            }
        }

        return low;
    }
    public int isPossible(int[] arr, int sum){
        int currArr = 1;
        int currSum = 0;
        
        for(int i = 0 ; i < arr.length; i++){
            if(currSum + arr[i] <= sum){
                currSum += arr[i];
            } else {
                currArr++;
                currSum = arr[i];
            }
        }

        return currArr;
    }
}

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.py
class Solution:
    def splitArray(self, nums, k):
        if not nums or len(nums) < k:
            return -1

        low, high = max(nums), sum(nums)

        while low <= high:
            mid = low + (high - low) // 2
            sub_arrays = self.isPossible(nums, mid)

            if sub_arrays > k:
                low = mid + 1
            else:
                high = mid - 1

        return low

    def isPossible(self, arr, target_sum):
        curr_arr = 1
        curr_sum = 0

        for num in arr:
            if curr_sum + num <= target_sum:
                curr_sum += num
            else:
                curr_arr += 1
                curr_sum = num

        return curr_arr
